### ISSUE TITLE
VM LinkedHashMap has inefficient `keys.contains`, likely others.

### ISSUE THREAD

**Issue by lrhn** [dart-lang/sdk#52909]
The `Map` protocol specifies that `map.keys.contains` should be as efficient as `map.containsKey`.

The current VM implementation performs a linear lookup on `keys.contains`, using the default `Iterable` implementation of `contains`. 

See: https://dartpad.dev/?id=180ce0b005f5345f9285147c222e5032 for a benchmark comparing `Map.containsKey` to `Map.keys.contains`.
Run in DartPad, dart2js shows results like:
```
map.containsKey: 98180.899 /ms (262143000 ops / 2670 ms)
map.keys.contains: 97705.181 /ms (262143000 ops / 2683 ms)
```
(with a variance bigger than the difference, so pretty much the same performance.)

The VM shows results like:
```
map.containsKey: 92076.923 /ms (262143000 ops / 2847 ms)
map.keys.contains: 15.294 /ms (31000 ops / 2027 ms)
```
Just a factor 6000 off. For a ~10000 element linear lookup, that's in the expected ballpark.

This is fairly easy to fix, it just needs to forward to `containsKey`. I'll likely do it myself.

I'll check if there are other missing efficient implementations. 

* Likely `last` *can* be wastly more efficient, but it isn't promised to be efficient to begin with. Still, some users may assume that it is efficient.
* The `entries` iterable uses the default implementation, which iterates keys, then does lookup on the key to get the value. That can be more efficient when the key/value is known to be adjacent in the data list.

**Comment by lrhn** [comment]
Yes, making the`keys` iterable's `contains` forward to the map's `containsKey` is the most likely fix here.

**Comment by iamsahilsonawane** [comment]
I understand, but wouldn't that hinder the functionality of the Iterable itself? 

Again just curious

**Comment by iamsahilsonawane** [comment]
I mean as a user, when I am accessing the iterable keys, I am for sure aware of it being an iterable and not map and consider using containsKey instead if required?

**Comment by lrhn** [comment]
I think `remove` now does cause compaction at 50% tombstones, but you can probably still keep `first` being O(n/2). We could optimize and cache the first possible non-tombstone, a "usedStart" like we have "usedSize" to remember the *last* place we need to look in the backing array. Whenever we iterate, and the "usedStart" is a tombstone, we remember the new first element.
Heck, we can make the data array be used as a cyclic buffer, so if you keep doing `set.add(set.removeFirst())`, you'll stay in the same array.

**Comment by a-siva** [comment]
@lrhn are you still considering fixing it? Does it need to be 'area-vm' issue or 'area-library' and is P2 the correct priority for it.

**Comment by lrhn** [comment]
Let's assume I won't be the one fixing it, since I haven't done so yet

**Comment by rakudrama** [comment]
The initial issue, that `.keys.contains` is slow for the VM default Map implementation, is fixed by 23a98d808c05f0b4a9ca0020e6162421ec08806f.
`HashMap` already has a `.keys.contains` that uses the Map's `containsKey`.

I am going to leave this issue open, since `.keys.contains` is still slow for `const` Maps.
